File: geocoder-googlemaps/src/main/java/com/prolificinteractive/geocoder/api/googlemaps/Parser.java
/*
 * Copyright (C) 2015 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.prolificinteractive.geocoder.api.googlemaps;

import android.support.annotation.NonNull;
import com.prolificinteractive.geocoder.model.Address;
import com.prolificinteractive.geocoder.model.Bounds;
import com.prolificinteractive.geocoder.model.Location;
import com.prolificinteractive.geocoder.model.Viewport;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Parser for Geocoder
 */
public final class Parser {

    /*
     * Status codes which we handle
     */

  private static final String ERROR_MESSAGE = "error_message";

  private static final String STATUS = "status";

  private static final String RESULTS = "results";

  private static final String GEOMETRY = "geometry";

  private static final String LOCATION = "location";

  private static final String LOCATION_TYPE = "location_type";

  private static final String VIEWPORT = "viewport";

  private static final String BOUNDS = "bounds";

  private static final String SOUTHWEST = "southwest";

  private static final String NORTHEAST = "northeast";

  private static final String LAT = "lat";

  private static final String LNG = "lng";

  private static final String FORMATTED_ADDRESS = "formatted_address";

  private static final String ADDRESS_COMPONENTS = "address_components";

  private static final String TYPES = "types";

  private static final String LONG_NAME = "long_name";

  private static final String SHORT_NAME = "short_name";

  private Parser() {
  }

  private static void parseAddressComponents(
      @NonNull final JSONObject result,
      @NonNull final Address.Builder addressBuilder) throws JSONException {
    if (result.has(ADDRESS_COMPONENTS)) {
      final JSONArray addressComponents = result
          .getJSONArray(ADDRESS_COMPONENTS);
      for (int a = 0; a < addressComponents.length(); a++) {
        final JSONObject addressComponent = addressComponents.getJSONObject(a);
        if (!addressComponent.has(TYPES)) {
          continue;
        }
        String value = null;
        if (addressComponent.has(LONG_NAME)) {
          value = addressComponent.getString(LONG_NAME);
        } else if (addressComponent.has(SHORT_NAME)) {
          value = addressComponent.getString(SHORT_NAME);
        }
        if (value == null || value.isEmpty()) {
          continue;
        }
        final JSONArray types = addressComponent.getJSONArray(TYPES);
        for (int t = 0; t < types.length(); t++) {
          final String type = types.getString(t);
          switch (type) {
            case "street_address":
              addressBuilder.streetAddress(value);
              break;

            case "route":
              addressBuilder.route(value);
              break;

            case "intersection":
              addressBuilder.intersection(value);
              break;

            case "political":
              addressBuilder.political(value);
              break;

            case "country":
              addressBuilder.country(value);
              break;

            case "administrative_area_level_1":
              addressBuilder.administrativeAreaLevel1(value);
              break;

            case "administrative_area_level_2":
              addressBuilder.administrativeAreaLevel2(value);
              break;

            case "administrative_area_level_3":
              addressBuilder.administrativeAreaLevel3(value);
              break;

            case "administrative_area_level_4":
              addressBuilder.administrativeAreaLevel4(value);
              break;

            case "administrative_area_level_5":
              addressBuilder.administrativeAreaLevel5(value);
              break;

            case "colloquial_area":
              addressBuilder.colloquialArea(value);
              break;

            case "locality":
              addressBuilder.locality(value);
              break;

            case "ward":
              addressBuilder.ward(value);
              break;

            case "sublocality":
              addressBuilder.subLocality(value);
              break;

            case "sublocality_level_1":
              addressBuilder.subLocalityLevel1(value);
              break;

            case "sublocality_level_2":
              addressBuilder.subLocalityLevel2(value);
              break;

            case "sublocality_level_3":
              addressBuilder.subLocalityLevel3(value);
              break;

            case "sublocality_level_4":
              addressBuilder.subLocalityLevel4(value);
              break;

            case "sublocality_level_5":
              addressBuilder.subLocalityLevel5(value);
              break;

            case "neighborhood":
              addressBuilder.neighborhood(value);

              break;
            case "premise":
              addressBuilder.premise(value);
              break;

            case "subpremise":
              addressBuilder.subPremise(value);
              break;

            case "postal_code":
              addressBuilder.postalCode(value);
              break;

            case "natural_feature":
              addressBuilder.naturalFeature(value);
              break;

            case "airport":
              addressBuilder.airport(value);
              break;

            case "park":
              addressBuilder.park(value);
              break;

            case "point_of_interest":
              addressBuilder.pointOfInterest(value);
              break;

            case "floor":
              addressBuilder.floor(value);
              break;

            case "establishment":
              addressBuilder.establishment(value);
              break;

            case "parking":
              addressBuilder.parking(value);
              break;

            case "post_box":
              addressBuilder.postBox(value);
              break;

            case "postal_town":
              addressBuilder.postTown(value);
              break;

            case "room":
              addressBuilder.room(value);
              break;

            case "street_number":
              addressBuilder.streetNumber(value);
              break;

            case "bus_station":
              addressBuilder.busStation(value);
              break;

            case "train_station":
              addressBuilder.trainStation(value);
              break;

            case "transit_station":
              addressBuilder.transitStation(value);
              break;

            default:
              // Unhandled
              break;
          }
        }
      }
    }
  }

  private static void parseGeometry(
      @NonNull final JSONObject result,
      @NonNull final Address.Builder builder) throws JSONException {
    if (result.has(GEOMETRY)) {
      final JSONObject geometry = result.getJSONObject(GEOMETRY);
      if (geometry.has(LOCATION_TYPE)) {
        builder.locationType(geometry.getString(LOCATION_TYPE));
      }

      if (geometry.has(LOCATION)) {
        final JSONObject location = geometry.getJSONObject(LOCATION);
        builder.location(Location.builder()
            .latitude(location.getDouble(LAT))
            .longitude(location.getDouble(LNG))
            .build());
      }

      if (geometry.has(VIEWPORT)) {
        final JSONObject viewport = geometry.getJSONObject(VIEWPORT);
        if (viewport.has(SOUTHWEST) && viewport.has(NORTHEAST)) {
          final JSONObject southwest = viewport.getJSONObject(SOUTHWEST);
          final Location locationSouthwest = Location.builder()
              .latitude(southwest.getDouble(LAT))
              .longitude(southwest.getDouble(LNG))
              .build();

          final JSONObject northeast = viewport.getJSONObject(NORTHEAST);
          final Location locationNortheast = Location.builder()
              .latitude(northeast.getDouble(LAT))
              .longitude(northeast.getDouble(LNG))
              .build();

          builder.viewport(Viewport.builder()
              .northeast(locationNortheast)
              .southwest(locationSouthwest)
              .build());
        }
      }

      if (geometry.has(BOUNDS)) {
        final JSONObject viewport = geometry.getJSONObject(BOUNDS);
        if (viewport.has(SOUTHWEST) && viewport.has(NORTHEAST)) {
          final JSONObject southwest = viewport.getJSONObject(SOUTHWEST);
          final Location locationSouthwest = Location.builder()
              .latitude(southwest.getDouble(LAT))
              .longitude(southwest.getDouble(LNG))
              .build();

          final JSONObject northeast = viewport.getJSONObject(NORTHEAST);
          final Location locationNortheast = Location.builder()
              .latitude(northeast.getDouble(LAT))
              .longitude(northeast.getDouble(LNG))
              .build();

          builder.bounds(Bounds.builder()
              .southwest(locationSouthwest)
              .northeast(locationNortheast)
              .build());
        }
      }
    }
  }

  @NonNull
  static List<Address> parseJson(
      final byte[] jsonData,
      final int maxResults,
      final boolean parseAddressComponents)
      throws Exception {
    try {
      final String jsonString = new String(jsonData, Charset.forName("UTF-8"));
      final JSONObject jsonObject = new JSONObject(jsonString);

      if (!jsonObject.has(STATUS)) {
        throw new Exception(new JSONException("No \"status\" field"));
      }

      final Status status = Status.fromString(jsonObject.getString(STATUS));
      switch (status) {
        case OK:
          if (jsonObject.has(RESULTS)) {
            return parseResults(maxResults, parseAddressComponents, jsonObject);
          }
          return new ArrayList<>();

        case ZERO_RESULTS:
          return new ArrayList<>();

        default:
          final GoogleMapsException e = GoogleMapsException.forStatus(status);
          try {
            if (jsonObject.has(ERROR_MESSAGE)) {
              e.setErrorMessage(jsonObject.getString(ERROR_MESSAGE));
            }
          } catch (final JSONException ignored) {
          }
          throw e;
      }
    } catch (final JSONException e) {
      throw new GoogleMapsException(e);
    }
  }

  private static List<Address> parseResults(
      final int maxResults,
      final boolean parseAddressComponents,
      @NonNull final JSONObject o) throws JSONException {
    final JSONArray results = o.getJSONArray(RESULTS);
    final int count = results.length() >= maxResults ? maxResults : results.length();
    final ArrayList<Address> addressList = new ArrayList<>(count);
    for (int i = 0; i < count; i++) {

      final Address.Builder addressBuilder = Address.builder();
      final JSONObject result = results.getJSONObject(i);

      if (result.has(FORMATTED_ADDRESS)) {
        addressBuilder.formattedAddress(result.getString(FORMATTED_ADDRESS));
      }

      parseGeometry(result, addressBuilder);

      if (parseAddressComponents) {
        parseAddressComponents(result, addressBuilder);
      }

      addressList.add(addressBuilder.build());
    }
    return addressList;
  }
}

File: geocoder-googlemaps/src/main/java/com/prolificinteractive/geocoder/api/googlemaps/Parser.kt
/*
 * Copyright (C) 2015 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.prolificinteractive.geocoder.api.googlemaps

import com.prolificinteractive.geocoder.model.Address
import com.prolificinteractive.geocoder.model.Bounds
import com.prolificinteractive.geocoder.model.Location
import com.prolificinteractive.geocoder.model.Viewport
import org.json.JSONException
import org.json.JSONObject
import java.nio.charset.Charset
import java.util.ArrayList

/**
 * Parser for Geocoder
 */
class Parser {

  companion object {

    /*
   * Status codes which we handle
   */

    private val ERROR_MESSAGE = "error_message"

    private val STATUS = "status"

    private val RESULTS = "results"

    private val GEOMETRY = "geometry"

    private val LOCATION = "location"

    private val LOCATION_TYPE = "location_type"

    private val VIEWPORT = "viewport"

    private val BOUNDS = "bounds"

    private val SOUTHWEST = "southwest"

    private val NORTHEAST = "northeast"

    private val LAT = "lat"

    private val LNG = "lng"

    private val FORMATTED_ADDRESS = "formatted_address"

    private val ADDRESS_COMPONENTS = "address_components"

    private val TYPES = "types"

    private val LONG_NAME = "long_name"

    private val SHORT_NAME = "short_name"


    @Throws(JSONException::class)
    private fun parseAddressComponents(
        result: JSONObject,
        address: Address): Address {

      var formattedAddress: String? = null
      var streetAddress: String? = null
      var route: String? = null
      var intersection: String? = null
      var political: String? = null
      var country: String? = null
      var administrativeAreaLevel1: String? = null
      var administrativeAreaLevel2: String? = null
      var administrativeAreaLevel3: String? = null
      var administrativeAreaLevel4: String? = null
      var administrativeAreaLevel5: String? = null
      var colloquialArea: String? = null
      var locality: String? = null
      var ward: String? = null
      var subLocality: String? = null
      var subLocalityLevel1: String? = null
      var subLocalityLevel2: String? = null
      var subLocalityLevel3: String? = null
      var subLocalityLevel4: String? = null
      var subLocalityLevel5: String? = null
      var neighborhood: String? = null
      var premise: String? = null
      var subPremise: String? = null
      var postalCode: String? = null
      var naturalFeature: String? = null
      var airport: String? = null
      var park: String? = null
      var pointOfInterest: String? = null
      var floor: String? = null
      var establishment: String? = null
      var parking: String? = null
      var postBox: String? = null
      var postTown: String? = null
      var room: String? = null
      var streetNumber: String? = null
      var busStation: String? = null
      var trainStation: String? = null
      var transitStation: String? = null

      if (result.has(ADDRESS_COMPONENTS)) {
        val addressComponents = result
            .getJSONArray(ADDRESS_COMPONENTS)
        for (a in 0 until addressComponents.length()) {
          val addressComponent = addressComponents.getJSONObject(a)
          if (!addressComponent.has(TYPES)) {
            continue
          }
          var value: String? = null
          if (addressComponent.has(LONG_NAME)) {
            value = addressComponent.getString(LONG_NAME)
          } else if (addressComponent.has(SHORT_NAME)) {
            value = addressComponent.getString(SHORT_NAME)
          }
          if (value == null || value.isEmpty()) {
            continue
          }
          val types = addressComponent.getJSONArray(TYPES)
          for (t in 0 until types.length()) {
            val type = types.getString(t)
            when (type) {
              "street_address" -> streetAddress = value
              "route" -> route = value
              "intersection" -> intersection = value
              "political" -> political = value
              "country" -> country = value
              "administrative_area_level_1" -> administrativeAreaLevel1 = value
              "administrative_area_level_2" -> administrativeAreaLevel2 = value
              "administrative_area_level_3" -> administrativeAreaLevel3 = value
              "administrative_area_level_4" -> administrativeAreaLevel4 = value
              "administrative_area_level_5" -> administrativeAreaLevel5 = value
              "colloquial_area" -> colloquialArea = value
              "locality" -> locality = value
              "ward" -> ward = value
              "sublocality" -> subLocality = value
              "sublocality_level_1" -> subLocalityLevel1 = value
              "sublocality_level_2" -> subLocalityLevel2 = value
              "sublocality_level_3" -> subLocalityLevel3 = value
              "sublocality_level_4" -> subLocalityLevel4 = value
              "sublocality_level_5" -> subLocalityLevel5 = value
              "neighborhood" -> neighborhood = value
              "premise" -> premise = value
              "subpremise" -> subPremise = value
              "postal_code" -> postalCode = value
              "natural_feature" -> naturalFeature = value
              "airport" -> airport = value
              "park" -> park = value
              "point_of_interest" -> pointOfInterest = value
              "floor" -> floor = value
              "establishment" -> establishment = value
              "parking" -> parking = value
              "post_box" -> postBox = value
              "postal_town" -> postTown = value
              "room" -> room = value
              "street_number" -> streetNumber = value
              "bus_station" -> busStation = value
              "train_station" -> trainStation = value
              "transit_station" -> transitStation = value
            }
          }
        }
      }

      return address.copy(formattedAddress = formattedAddress,
          streetAddress = streetAddress,
          route = route,
          intersection = intersection,
          political = political,
          country = country,
          administrativeAreaLevel1 = administrativeAreaLevel1,
          administrativeAreaLevel2 = administrativeAreaLevel2,
          administrativeAreaLevel3 = administrativeAreaLevel3,
          administrativeAreaLevel4 = administrativeAreaLevel4,
          administrativeAreaLevel5 = administrativeAreaLevel5,
          colloquialArea = colloquialArea,
          locality = locality,
          ward = ward,
          subLocality = subLocality,
          subLocalityLevel1 = subLocalityLevel1,
          subLocalityLevel2 = subLocalityLevel2,
          subLocalityLevel3 = subLocalityLevel3,
          subLocalityLevel4 = subLocalityLevel4,
          subLocalityLevel5 = subLocalityLevel5,
          neighborhood = neighborhood,
          premise = premise,
          subPremise = subPremise,
          postalCode = postalCode,
          naturalFeature = naturalFeature,
          airport = airport,
          park = park,
          pointOfInterest = pointOfInterest,
          floor = floor,
          establishment = establishment,
          parking = parking,
          postBox = postBox,
          postTown = postTown,
          room = room,
          streetNumber = streetNumber,
          busStation = busStation,
          trainStation = trainStation,
          transitStation = transitStation)
    }

    private fun parseGeometry(
        result: JSONObject,
        address: Address): Address {

      var locationType: String? = null
      var location: Location? = null
      var viewport: Viewport? = null
      var bounds: Bounds? = null

      if (result.has(GEOMETRY)) {
        val geometry = result.getJSONObject(GEOMETRY)
        if (geometry.has(LOCATION_TYPE)) {
          locationType = geometry.getString(LOCATION_TYPE)
        }

        if (geometry.has(LOCATION)) {
          val locationJson = geometry.getJSONObject(LOCATION)
          location = Location(locationJson.getDouble(LAT), locationJson.getDouble(LNG))
        }

        if (geometry.has(VIEWPORT)) {
          val viewportJson = geometry.getJSONObject(VIEWPORT)
          if (viewportJson.has(SOUTHWEST) && viewportJson.has(NORTHEAST)) {
            val southwest = viewportJson.getJSONObject(SOUTHWEST)
            val locationSouthwest = Location(southwest.getDouble(LAT), southwest.getDouble(LNG))

            val northeast = viewportJson.getJSONObject(NORTHEAST)
            val locationNortheast = Location(northeast.getDouble(LAT), northeast.getDouble(LNG))

            viewport = Viewport(locationNortheast, locationSouthwest)
          }
        }

        if (geometry.has(BOUNDS)) {
          val viewportJson = geometry.getJSONObject(BOUNDS)
          if (viewportJson.has(SOUTHWEST) && viewportJson.has(NORTHEAST)) {
            val southwest = viewportJson.getJSONObject(SOUTHWEST)
            val locationSouthwest = Location(southwest.getDouble(LAT), southwest.getDouble(LNG))

            val northeast = viewportJson.getJSONObject(NORTHEAST)
            val locationNortheast = Location(northeast.getDouble(LAT) ,(northeast.getDouble(LNG)))

            bounds = Bounds(locationSouthwest, locationNortheast)
          }
        }
      }

      return address.copy(
          locationType = locationType,
          location = location,
          viewport = viewport,
          bounds = bounds
      )
    }

    @Throws(GoogleMapsException::class)
    fun parseJson(
        jsonData: ByteArray,
        maxResults: Int,
        parseAddressComponents: Boolean): List<Address> {
      try {
        val jsonString = String(jsonData, Charset.forName("UTF-8"))
        val jsonObject = JSONObject(jsonString)

        if (!jsonObject.has(STATUS)) {
          throw Exception(JSONException("No \"status\" field"))
        }

        val status = Status.fromString(jsonObject.getString(STATUS))
        when (status) {
          Status.OK -> {
            return if (jsonObject.has(RESULTS)) {
              parseResults(maxResults, parseAddressComponents, jsonObject)
            } else ArrayList()
          }

          Status.ZERO_RESULTS -> return ArrayList()

          else -> {
            val e = GoogleMapsException.forStatus(status)
            try {
              if (jsonObject.has(ERROR_MESSAGE)) {
                e.errorMessage = jsonObject.getString(ERROR_MESSAGE)
              }
            } catch (ignored: JSONException) {
            }

            throw e
          }
        }
      } catch (e: JSONException) {
        throw GoogleMapsException(e)
      }

    }

    @Throws(JSONException::class)
    fun parseResults(
        maxResults: Int,
        parseAddressComponents: Boolean,
        jsonObject: JSONObject): List<Address> {
      val results = jsonObject.getJSONArray(RESULTS)
      val count = if (results.length() >= maxResults && maxResults != -1) maxResults else results.length()
      val addressList = ArrayList<Address>(count)
      for (i in 0 until count) {

        var address = Address()
        val result = results.getJSONObject(i)

        if (result.has(FORMATTED_ADDRESS)) {
          address = address.copy(formattedAddress = result.getString(FORMATTED_ADDRESS))
        }

        address = parseGeometry(result, address)

        if (parseAddressComponents) {
          address = parseAddressComponents(result, address)
        }

        addressList.add(address)
      }
      return addressList
    }
  }

}
